src/core/lp_parser.mly

```diff
@@ -0,0 +1,379 @@
+    {
+      let t =
+        if ps = [] then t else
+        make_pos ($startpos(ps), $endpos(a)) (P_Abst(ps, t))
+      in
+      let a =
+        if ps = [] then a else
+        make_pos ($startpos(ps), $endpos(a)) (P_Prod(ps, a))
+      in
+      make_pos $loc (P_query_assert(k, P_assert_typing(t, a)))
+    }
+  | k=assert_not ps=arg* TURNSTILE t=term EQUIV a=term
+    {
+      let t =
+        if ps = [] then t else
+        make_pos ($startpos(ps), $endpos(a)) (P_Abst(ps, t))
+      in
+      let a =
+        if ps = [] then a else
+        make_pos ($startpos(ps), $endpos(a)) (P_Abst(ps, a))
+      in
+      make_pos $loc (P_query_assert(k, P_assert_conv(t, a)))
+    }
+
+// Top level commands
+command:
+  | REQUIRE OPEN p=path+ SEMICOLON { make_pos $loc (P_require(true,p)) }
+  | REQUIRE p=path+ SEMICOLON { make_pos $loc (P_require(false, p)) }
+  | REQUIRE p=path AS a=ident SEMICOLON
+      {
+        let alias = Pos.make (fst a).pos ((fst a).elt, snd a) in
+        make_pos $loc (P_require_as(p, alias))
+      }
+  | OPEN p=path SEMICOLON { make_pos $loc (P_open([p])) }
+  | ms=modifier* SYMBOL s=ident al=arg* COLON a=term SEMICOLON
+      { make_pos $loc (P_symbol(ms, fst s, al, a)) }
+  | ms=modifier* INDUCTIVE i=any_ident COLON a=term ASSIGN VBAR?
+    c=separated_list(VBAR, separated_pair(any_ident, COLON, term))
+    SEMICOLON
+      { make_pos $loc (P_inductive(ms, i, a, c)) }
+  | ms=modifier* DEFINITION s=ident al=arg* ao=preceded(COLON, term)?
+    ASSIGN b=term SEMICOLON
+      { make_pos $loc (P_definition(ms, false, fst s, al, ao, b)) }
+  | ms=modifier* THEOREM s=ident al=arg* COLON a=term
+    PROOF ts=ttactic* pe=PROOF_END
+      {
+        let st = make_pos ($startpos(s), $endpos(a)) (fst s, al, a) in
+        let pe = make_pos $loc(pe) pe in
+        make_pos $loc (P_theorem(ms, st, ts, pe)) }
+  | RULE rs=separated_nonempty_list(WITH, rule) SEMICOLON
+      { make_pos $loc (P_rules(rs)) }
+  | SET c=config SEMICOLON { make_pos $loc (P_set(c)) }
+  | q=query SEMICOLON { make_pos $loc (P_query(q)) }
+  | EOF { raise End_of_file }
+
+// Environment of a metavariable or rewrite pattern
+env: L_SQ_BRACKET ts=separated_list(SEMICOLON, term) R_SQ_BRACKET { ts }
+
+sterm:
+  // Explicited identifier (with @)
+  | AT qid=qident { make_pos $loc (P_Iden(qid, true)) }
+  // Qualified identifier
+  | qid=qident { make_pos $loc (P_Iden(qid, false)) }
+  // The wildcard "_"
+  | WILD { make_pos $loc P_Wild }
+  // The constant [TYPE] (of type Kind)
+  | TYPE { make_pos $loc P_Type }
+  // Metavariable
+  | QUESTION_MARK m=ident e=env?
+      { make_pos $loc (P_Meta(fst m, Option.map Array.of_list e)) }
+  // Pattern of rewrite rules
+  | p=patt e=env? { make_pos $loc (P_Patt(p, Option.map Array.of_list e)) }
+  // Parentheses
+  | L_PAREN t=term R_PAREN { make_pos $loc (P_Wrap(t)) }
+  // Explicitly given argument
+  | L_CU_BRACKET t=term R_CU_BRACKET { make_pos $loc (P_Expl(t)) }
```

Comment:
What do you think to using square brackets for implicit arguments instead, and reserve curly brackets for (later) denoting subtypes?

Reply:
We can try, and keep it if it doesn't introduce ambiguity

Reply:
For unification rules, we could use begin/end instead of square brackets?

Reply:
We could indeed.

Reply:
Using square brackets for implicit arguments introduce an ambiguity with enironments and rewrite patterns, so we may keep the curly brackets (but we can use `begin` `end` for unification rules)

Reply:
ok